WEB: Announce Director update (#357)
scummvm/scummvm-web: data/en/news/20231209.md

---
title: End of year Director roundup
author: moralrecordings
date: 1702127996
---

> The end of the year is almost upon us, along with a new release of ScummVM. So I figured I'd provide a quick update for one of the more ambitious sub-projects that has been percolating along in the background: Macromedia Director support! For those who aren't aware, Macromedia Director was a popular framework used for thousands of games, educational titles, catalogues, art pieces, and other multimedia works.

Reviewer: > So I figured I'd provide

Past ScummVM announcements have been written from the point of view of the ScummVM team as a whole, rather than an individual. You can either change "I" to "we" (as you did for line 12) or remove the need for any first-person pronouns.

```suggestion
The end of the year is almost upon us, along with a new release of ScummVM that includes one of the more ambitious sub-projects that has been percolating along in the background: Macromedia Director support! For those who aren't aware, Macromedia Director was a popular framework used for thousands of games, educational titles, catalogues, art pieces, and other multimedia works.
```

New Games For Testing
=====================

First off, the upcoming 2.8.0 release will have some new supported games, and we need your help testing them!

> - Alice: An Interactive Museum

Author: Fixed

- Ape Odyssey
- Classical Cats
- Gadget: Invention, Travel and Adventure
- The Cute Machine
- The Dark Eye
- Wrath of the Gods

> You will need the data files from the games' original install media, and a [daily build](https://www.scummvm.org/downloads/#daily) of ScummVM. In the case of The Dark Eye, you will also need the files created by the installer. As always, if you run into issues, please let us know on the [issue tracker](https://bugs.scummvm.org/).

Author: Good point, added a Dumper Companion note. I'll try and get some wiki stubs ready after I get the code PRs for 2.8.0 lined up.

> If your fave isn't in the supported list, don't despair! ScummVM can open all Director titles from 1995 and earlier, so grab the latest [daily build](https://www.scummvm.org/downloads/#daily) and give it a try. 

Reviewer 2: Shall we clarify somehow that the launch is not guaranteed and that they could reach out to the team and ask for support?

Author: Had another go at the wording for this to include caveat + support link

> If your fave isn't in the supported list, don't despair! ScummVM can open all Director titles from 1995 and earlier, so grab the latest [daily build](https://www.scummvm.org/downloads/#daily) and give it a try. 

Reviewer: "Fave" is too colloquial and the translators to other languages might have a hard time with it.
```suggestion
If your favorite game isn't in the supported list, don't despair! ScummVM can open all Director titles from 1995 and earlier, so grab the latest [daily build](https://www.scummvm.org/downloads/#daily) and give it a try. 
```

Author: I've changed it to favourite... given Director is used in so many non-"game" things I've been trying to stick to "title"

> If you've encountered a title based on Director that isn't recognised, that's great! Please let us know on the [issue tracker](https://bugs.scummvm.org/) so it can be added.

Reviewer: ScummVM news announcements use American English.
```suggestion
If you've encountered a title based on Director that isn't recognized, that's great! Please let us know on the [issue tracker](https://bugs.scummvm.org/) so it can be added.
```

Engine versions 
===============

> As of the time of writing, there are **1595** titles in the Director detection list, with the following counts by Director version:

Reviewer: Those less familiar with ScummVM innards may not understand this part. Can we simplify it like this?

```suggestion
As of the time of writing, ScummVM can detect **1595** different Director titles, with the following counts by Director version:
```

Author: Fixed

| D0 | D1 | D2 | D3  | D4  | D5  | D6  | D7 | D8  | D9 | D10 | D11 |
|----|----|----|-----|-----|-----|-----|----|-----|----|-----|-----|
| 12 | 1  | 14 | 288 | 602 | 262 | 194 | 97 | 138 | 31 | 36  | 4   |

> ScummVM 2.8.0 will be the first release to include preliminary support for D5! A lot of work has gone into overhauling the codebase to support multiple cast libraries, however there is still plenty of work to be done adding Lingo functions and Xtras.

Reviewer: > preliminary support for D5

Is "D5" an official name? If so, it's fine, otherwise it might be good to spell it out as "version 5 of Director" or something.

Author: Fair, it's a bit confusing. In the code + documentation + issue tracker the standard is to use D[major version] to reference the version of the Director engine. I changed it so that the text refers to e.g. "D4 version of Director"

Reviewer: Works for me, thanks!

> ScummVM 2.8.0 will be the first release to include preliminary support for D5! A lot of work has gone into overhauling the codebase to support multiple cast libraries, however there is still plenty of work to be done adding Lingo functions and Xtras.

Reviewer: > multiple cast libraries, however there is still plenty of work to be done adding Lingo functions and Xtras.

This and everything afterwards (including the "XLibs" section) are full of technical jargon that may not be of interest to non-developers. If you think this is useful, it can be kept, but I think we might be better off cutting it.

Author: From the Discord discussion I think people are happy to see how this goes with the extra detail

Reviewer: I'll go with whatever the team consensus is.

Reviewer 3: I also agree with @Thunderforge that the announcement for the Director support is not the place to discuss too many technical details. The technical part could be placed in places like:
- the engine's wiki page
- a forum post
- a director-oriented blog

The news section contains "short and sweet" information about new and shiny stuff. Filling it with too many technical details is not ideal, as this section is mostly followed by gaming sites and people who are not worried too much about the technical details.

Reviewer 2: I liked the post.

Adding support for a new Director version is a tricky process. First we need to check for any changes to the movie file format, then add in any new engine and Lingo features, and finally start working through issues in a bunch of candidate games. Because of the difficulty of keeping track of facts vs. assumptions, we only enable features for an engine version once they have been confirmed with testing. This is why D6 and higher games will not open at all, even if the code we have would somewhat work.

> Support for D2, D3 and D4 continues to go from strength to strength. We are slowly building up a suite of unit tests, both for Lingo and for the various drawing modes in the renderer. Most of the remaining bugs are related to undefined or incidental behaviour: aspects like how two types in Lingo are compared, or the precise order that events are processed in, or how functions respond when you pass them nonsense data? All of these things are relied upon by thousands of Director products! And ScummVM must support all of them as closely as possible.

Reviewer: > continues to go from strength to strength

Cambridge Dictionary [says](https://dictionary.cambridge.org/dictionary/english/go-from-strength-to-strength) this is a UK expression. I had never heard of it as a native American speaker, and it would probably be difficult for translators. I think changing to "continues to improve" would be better.

Author: Replaced

> Support for D2, D3 and D4 continues to go from strength to strength. We are slowly building up a suite of unit tests, both for Lingo and for the various drawing modes in the renderer. Most of the remaining bugs are related to undefined or incidental behaviour: aspects like how two types in Lingo are compared, or the precise order that events are processed in, or how functions respond when you pass them nonsense data? All of these things are relied upon by thousands of Director products! And ScummVM must support all of them as closely as possible.

Reviewer: > behaviour

As before: use American English ("behavior").

Author: Fixed

XLibs
=====

Macromedia provided a C API with Director, which proved to be very popular for adding new engine features. There are hundreds of XObject and Xtra libraries in circulation, and each has to be reverse engineered and added as code to ScummVM. As of the time of writing, we have added support for **63** of these libraries.

A common use case is video; Director includes the ability to play back QuickTime videos as a sprite, but computers of the day would struggle with the added overhead of Director's renderer if the video resolution was larger than a postage stamp. To work around this, several XLibs exist to play back video as an accelerated overlay. Quite a few D3 and D4 titles are missing video playback due to this type of XLib, and we plan on creating a common method of supporting these.

> The good news is that most uncommon XLibs did only one or two things, and are pretty wasy to add initial support for. Part of the XLib format is a list of docstrings explaining each of the functions; there is a script in `/devtools` which can turn these docstrings into stubbed C++ code, saving a lot of hassle. 

Reviewer: ```suggestion
The good news is that most uncommon XLibs did only one or two things, and are pretty easy to add initial support for. Part of the XLib format is a list of docstrings explaining each of the functions; there is a script in `/devtools` which can turn these docstrings into stubbed C++ code, saving a lot of hassle. 
```

Author: Fixed

-----

> That's about it! We have some very dedicated volunteers scouring every swapmeet and bargain bin for new Director titles and testing them; if that sounds like your bag, you can visit them in the #engine-director channel of the [ScummVM Discord](https://discord.gg/4cDsMNtcpG). Until next time!

Reviewer: > scouring every swapmeet

We don't use that term in America and instead call them "flea markets". Besides, I don't think that's accurate because most people are probably searching eBay and whatever these days.

How about

```suggestion
That's about it! We have some very dedicated volunteers scouring for obscure Director titles and testing them; if that sounds like your bag, you can visit them in the `#engine-director` channel of the [ScummVM Discord](https://discord.gg/4cDsMNtcpG). Until next time!
```
